firedrake/functionspaceimpl.py

```diff
@@ -493,29 +493,38 @@ def exterior_facet_boundary_node_map(self, method):
         """
         return self._shared_data.exterior_facet_boundary_node_map(self, method)
 
+    def boundary_nodes(self, sub_domain, method):
+        """Return the boundary nodes for this :class:`~.FunctionSpace`.
+
+        :arg sub_domain: the mesh marker selecting which subset of facets to consider.
+        :arg method: the method for determining boundary nodes.
+        :returns: A numpy array of the unique function space nodes on
+           the selected portion of the boundary.
+
+        See also :class:`~.DirichletBC` for details of the arguments.
+        """
+        return self._shared_data.boundary_nodes(self, sub_domain, method)
```

Comment:
Ah, so we pass `self` (the function space) to the function space data object, which then can call back to `self.bottom_nodes` or `self.top_nodes` for example. Which in turn calls back to `top_bottom_boundary_nodes` on function space data. Why is this back and forth control necessary?